magenta-lib/src/main/scala/magenta/graph/Graph.scala

```diff
@@ -0,0 +1,101 @@
+package magenta.graph
+
+sealed trait Node[+T] {
+  def ~>[R >: T](to: Node[R]): Edge[R] = Edge(this, to)
+  def maybePriority: Option[Int] = None
+  def maybeValue: Option[T] = None
+}
+case object StartNode extends Node[Nothing]
+case class MidNode[T](value: T, priority: Int = 1) extends Node[T] {
+  override def maybePriority: Option[Int] = Some(priority)
+  def incPriority(n: Int): MidNode[T] = this.copy(priority = priority + n)
+  override def maybeValue: Option[T] = Some(value)
+}
+case object EndNode extends Node[Nothing]
+
+case class Edge[+T](from: Node[T], to: Node[T])
+
+object Graph {
+  def apply[T](edges: Edge[T]*): Graph[T] = Graph(edges.toSet)
+
+  def empty[T] = Graph[T](StartNode ~> EndNode)
+}
+
+case class Graph[T](edges: Set[Edge[T]]) {
+  val nodes = edges.map(_.from) ++ edges.map(_.to)
+  val dataNodes = nodes.filterMidNodes
+
+  val isValid: Either[List[String], Boolean] = {
+    val errors =
+      ((if (!nodes.contains(StartNode)) Some("No start node") else None) ::
+        (if (!nodes.contains(EndNode)) Some("No end node") else None) :: Nil).flatten
+    if (errors.isEmpty) Right(true) else Left(errors)
+  }
+  assert(isValid.isRight, s"Graph isn't valid: ${isValid.left.get.mkString(", ")}")
+
+  def get(node: T): MidNode[T] = dataNodes.find(_.value == node).get
+  def successors(node: Node[T]): Set[Node[T]] = edges.filter(_.from == node).map(_.to)
+  def predecessors(node: Node[T]): Set[Node[T]] = edges.filter(_.to == node).map(_.from)
+  def successorNodes(node: Node[T]): List[MidNode[T]] = {
```

Comment:
The idea is that it returns the successors in order of node priority.

In my recent restructure (to move priority back onto the edges - which makes it a lot easier to merge graphs as the nodes are comparable) I've also taken the opportunity to rename the method to orderedSuccessors.